### ISSUE TITLE
Columns with default value are updated even when they are unchanged

### ISSUE THREAD

**Issue by twinni** [yiisoft/yii2#8561]
Attributes in ActiveRecord models with default rules defined are updated on save() even when their value has not been changed.

```php
 public function rules()
    {
        return [
            [['ableseartID'], 'default', 'value' => 0],
            [['ueberlauf'], 'default', 'value' => false],
        ];
   }
```
This is the generated sql command when no field has been changed. 
```
UPDATE `zaehlerstand` SET `ueberlauf`=0, `ableseartID`=6 WHERE `id`=1023
```

**Comment by fbau123** [comment]
```php
    public function getDirtyAttributes($names = null)
    {
        if ($names === null) {
            $names = $this->attributes();
        }
        $names = array_flip($names);
        $attributes = [];
        if ($this->_oldAttributes === null) {
            foreach ($this->_attributes as $name => $value) {
                if (isset($names[$name])) {
                    $attributes[$name] = $value;
                }
            }
        } else {
            foreach ($this->_attributes as $name => $value) {
                if (isset($names[$name]) && (!array_key_exists($name, $this->_oldAttributes) || $value !== $this->_oldAttributes[$name])) {
                    $attributes[$name] = $value;
                }
            }
        }
        return $attributes;
    }
```
That is the method that finds all the attributes that should be updated, you have to make sure the new value is strictly equal to the old one.

**Comment by twinni** [comment]
@fbau123, thanks for your answer. You're right. All post data are returned as strings. Compared to the _oldAttributes value they are not equal.
But shouldn't there be a typecasting to the column type defined in the database **before** the call of  getDirtyAttributes()? 
If this would not be the case then all numeric fields will updated as soon as they are detected as safe!

**Comment by cebe** [comment]
can you show the table schema? also which DBMS are you using?

**Comment by twinni** [comment]
@cebe, here is the table definition. 
```
CREATE TABLE `zaehlerstand` (
	`id` INT(11) NOT NULL AUTO_INCREMENT,
	`zaehlerID` INT(11) NOT NULL,
	`ablesedatum` DATE NOT NULL,
	`ablesewert` DOUBLE NOT NULL,
	`ueberlauf` TINYINT(1) NOT NULL DEFAULT '0' ,
	`abrechdatum` DATE NULL DEFAULT NULL,
	`abrechwert` DOUBLE NULL DEFAULT NULL,
	`ableseartID` TINYINT(1) NOT NULL DEFAULT '0' ,
	`differenz` DOUBLE NULL DEFAULT NULL,
	`created_at` DATETIME NULL DEFAULT CURRENT_TIMESTAMP,
	`updated_at` DATETIME NULL DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
	`created_by` INT(11) NULL DEFAULT NULL,
	`updated_by` INT(11) NULL DEFAULT NULL,
	PRIMARY KEY (`id`),
	UNIQUE INDEX `secKey` (`zaehlerID`, `ablesedatum`),
	UNIQUE INDEX `stichtag_UNIQUE` (`zaehlerID`, `abrechdatum`),
	INDEX `fk_tblzaehlerstand_tblzaehler1_idx` (`zaehlerID`),
	INDEX `fk_tblzaehlerstand_tblrefableseart1_idx` (`ableseartID`),
	INDEX `fk_zaehlerstand_user1_idx` (`created_by`),
	INDEX `fk_zaehlerstand_user2_idx` (`updated_by`),
	CONSTRAINT `fk_zaehlerstand_ableseart1` FOREIGN KEY (`ableseartID`) REFERENCES `ableseart` (`id`) ON UPDATE CASCADE ON DELETE NO ACTION,
	CONSTRAINT `fk_zaehlerstand_user1` FOREIGN KEY (`created_by`) REFERENCES `user` (`id`) ON UPDATE CASCADE ON DELETE NO ACTION,
	CONSTRAINT `fk_zaehlerstand_user2` FOREIGN KEY (`updated_by`) REFERENCES `user` (`id`) ON UPDATE CASCADE ON DELETE NO ACTION,
	CONSTRAINT `fk_zaehlerstand_zaehler1` FOREIGN KEY (`zaehlerID`) REFERENCES `zaehler` (`id`) ON UPDATE CASCADE ON DELETE CASCADE
)
COLLATE='utf8_general_ci'
ENGINE=InnoDB
AUTO_INCREMENT=1052
;
```
The problem described above is not only true for default rules.
With the exception of int() columns no typcasting will happen so that all numeric, non int() columns, will be handled as dirty attributes.

I'm using the latest MySQL Server (5.6.24)

**Comment by chervand** [comment]
So, is it possible to avoid unnecessary SQL UPDATE request without explicit type casting of POST data?

**Comment by klimov-paul** [comment]
Resolved by #12067
